### PR TITLE
RDConfig: Fetch all AciVrf related (BD/EPGs) subnets

### PR DESCRIPTION
**Pull request by kishorejonnala**
- Build the subnet-cache during acc bootup
            - This will include all BD/EPG subnets corresponding to AciVrf
    - Subscribe to fvSubnet class with a filter of AviVrfRelated Tenants
    - Upon websocket notification, check and update subnet-cache as needed.
    - Also, trigger syncRdConfig (host-agent to consume) on subnet-cache change.

### CODE REVIEW

File: pkg/apicapi/apicapi.go [hunk 1]
```diff
@@ -1096,6 +1097,37 @@ func (conn *ApicConnection) SetSubscriptionHooks(value string,
...
+		conn.logErrorResp("Could not get subtree for "+url, resp)
+		return apicresp,err
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&apicresp)
+	if err != nil {
+		conn.log.Error("Could not parse APIC response: ", err)
+		return apicresp,err
+	}
+	complete(resp)
```
Comment (jayaramsatya) [inline pkg/apicapi/apicapi.go]: do we need this you called above defer call for this

File: pkg/controller/subnetcache.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			cont.log.Debug("subnetParentDn is in cachedSubnets:", subnetParentDn)
+			cont.apicConn.CachedSubnetDns[subnetDn] = subnetIp
+			inCache = true
+			cont.log.Debug("cachedSubnetDns Map:  ", cont.apicConn.CachedSubnetDns)
+			continue
+		}
+	}
+	if !inCache {
+		cont.BuildSubnetDnCache(subnetDn, aciVrfDn)
+		cont.triggerSyncRdConfig()
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go]: replace this with scheduleRdConfig

File: pkg/controller/subnetcache.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	if _, ok := cont.apicConn.CachedSubnetDns[dn]; ok {
+		delete(cont.apicConn.CachedSubnetDns, dn)
+	} else {
+		cont.log.Debug("This shouldn't happen!")
+	}
+	cont.log.Debug("After Delete: CachedSubnets.subnetDns Map:  ", cont.apicConn.CachedSubnetDns)
+	cont.scheduleRdConfig()
+}
+
+func (cont *AciController) triggerSyncRdConfig() {
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go]: delete this dummy call

File: pkg/controller/subnetcache.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        for _, body := range obj {
+			dn,ok := body.Attributes["dn"].(string)
+			if !ok {
+				continue
+			}
+			vrfBdDns = append(vrfBdDns, strings.Split(dn, bdRsDelimiter)[0] )
+		}
+    }
+    sort.Strings(vrfBdDns)	
+	cont.log.Debug("aciVrfBdDns: ", vrfBdDns)
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go]: nit: check indentation here

File: pkg/controller/subnetcache.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    "github.com/noironetworks/aci-containers/pkg/apicapi"
+)
+
+func (cont *AciController) SubnetChanged(obj apicapi.ApicObject, aciVrfDn string) {
+	subnetDn := obj.GetAttrStr("dn")
+	subnetIp := obj.GetAttrStr("ip")
+	
+	cont.log.Debug("SubnetChanged for dn: ", subnetDn)
+	if _, ok := cont.apicConn.CachedSubnetDns[subnetDn]; !ok {
+		cont.UpdateSubnetDnCache(subnetDn, subnetIp, aciVrfDn)
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go]: is there any possibility that change notification will called even though there is no change in subnets? in that case you can avoid calling the sync

File: pkg/controller/subnetcache.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+		cont.scheduleRdConfig()
+	}
+}
+
+func (cont *AciController) SubnetDeleted(dn string) {
+	cont.log.Debug("Before deleting dn: ", dn, " CachedSubnets.subnetDns Map:  ", cont.apicConn.CachedSubnetDns)
+	
+	if _, ok := cont.apicConn.CachedSubnetDns[dn]; ok {
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go]: if you are not deleting from the cache can we avoid calling sync

File: pkg/controller/subnetcache.go (hunk 2 above)
Comment (kishorejonnala) [reply pkg/controller/subnetcache.go]: Done.

File: pkg/apicapi/apicapi.go (hunk 1 above)
Comment (kishorejonnala) [reply pkg/apicapi/apicapi.go]: Oversight; removed.

File: pkg/controller/subnetcache.go (hunk 4 above)
Comment (kishorejonnala) [reply pkg/controller/subnetcache.go]: ran go fmt; you will see many other indentation changes now.

File: pkg/controller/subnetcache.go (hunk 6 above)
Comment (kishorejonnala) [reply pkg/controller/subnetcache.go]: Good one.. Fixed.

File: pkg/controller/subnetcache.go (hunk 5 above)
Comment (kishorejonnala) [reply pkg/controller/subnetcache.go]: Fixed.

**Pull request merged**